### PR TITLE
Get index includes parent data stream for backing indices

### PR DESCRIPTION
**Pull request by danhermann**
The `GET <index>` endpoint now includes the name of the parent data stream for any backing indices.

Relates to #53100.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java [hunk 1]
```diff
@@ -126,6 +131,13 @@ public GetIndexResponse(String[] indices,
             defaultSettingsMapBuilder.put(in.readString(), Settings.readSettingsFromStream(in));
         }
         defaultSettings = defaultSettingsMapBuilder.build();
+
+        ImmutableOpenMap.Builder<String, String> dataStreamsMapBuilder = ImmutableOpenMap.builder();
+        int dataStreamsSize = in.readVInt();
+        for (int i = 0; i < dataStreamsSize; i++) {
+            dataStreamsMapBuilder.put(in.readString(), in.readOptionalString());
+        }
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java:139]: I think this de-serialization logic needs to be inside a version check.
I think this is why the mixed cluster tests have failed.

File: server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java [hunk 2]
```diff
@@ -233,6 +253,11 @@ public void writeTo(StreamOutput out) throws IOException {
             out.writeString(indexEntry.key);
             Settings.writeSettingsToStream(indexEntry.value, out);
         }
+        out.writeVInt(dataStreams.size());
+        for (ObjectObjectCursor<String, String> indexEntry : dataStreams) {
+            out.writeString(indexEntry.key);
+            out.writeOptionalString(indexEntry.value);
+        }
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java:260]: Same applies for this serialization logic.

**Review by martijnvg — COMMENTED** [review]
This change looks good. I left a comment around bwc (de-)serialization logic.

**Comment by danhermann** [comment]
@elasticmachine update branch

**Review by martijnvg — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 3]
```diff
@@ -401,6 +401,24 @@ public boolean equalsAliases(Metadata other) {
         return indexMapBuilder.build();
     }
 
+    /**
+     * Finds the parent data streams, if any, for the specified concrete indices.
+     */
+    public ImmutableOpenMap<String, IndexAbstraction.DataStream> findDataStreams(String[] concreteIndices) {
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:407]: I wonder if this should not be a private method on `TransportGetIndexAction` instead? Seems it mainly translates from the lookup to the map structure.

**Review by henningandersen — APPROVED** [review]
LGTM.

**Pull request merged**